python/ray/includes/gcs_client.pxi

```diff
@@ -360,3 +488,151 @@ cdef class NewGcsClient:
+# to C++ for Async APIs.
+#
+# Each function accepts what the C++ callback passes, typically a Status and a value.
+# Returns `Tuple[object, Optional[Exception]]` (we are all gophers now lol).
+# Must not raise exceptions, or it crashes the process.
+#############################################################
+
+cdef convert_get_all_node_info(
+        CRayStatus status, c_vector[CGcsNodeInfo]&& c_data):
+    # -> Dict[JobID, gcs_pb2.JobTableData]
+    cdef c_string b
+    try:
+        check_status_timeout_as_rpc_error(status)
+        node_table_data = {}
+        for c_proto in c_data:
+            b = c_proto.SerializeAsString()
+            proto = gcs_pb2.GcsNodeInfo()
+            proto.ParseFromString(b)
+            node_table_data[proto.node_id] = proto
+        return node_table_data, None
+    except Exception as e:
+        return None, e
+
+cdef convert_get_all_job_info(
+        CRayStatus status, c_vector[CJobTableData]&& c_data):
+    # -> Dict[JobID, gcs_pb2.JobTableData]
+    cdef c_string b
+    try:
+        check_status_timeout_as_rpc_error(status)
+        job_table_data = {}
+        for c_proto in c_data:
+            b = c_proto.SerializeAsString()
+            proto = gcs_pb2.JobTableData()
+            proto.ParseFromString(b)
+            job_table_data[proto.job_id] = proto
+        return job_table_data, None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_str_none_for_not_found(
+        CRayStatus status, const optional[c_string]& c_str):
+    # If status is NotFound, return None.
+    # If status is OK, return the value.
+    # Else, raise exception.
+    # -> Optional[bytes]
+    try:
+        if status.IsNotFound():
+            return None, None
+        check_status_timeout_as_rpc_error(status)
+        return c_str.value(), None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_multi_get(
+        CRayStatus status, const optional[unordered_map[c_string, c_string]]& c_map):
+    # -> Dict[str, str]
+    cdef unordered_map[c_string, c_string].const_iterator it
+    try:
+        check_status_timeout_as_rpc_error(status)
+        result = {}
+        it = c_map.value().const_begin()
+        while it != c_map.value().const_end():
+            key = dereference(it).first
+            value = dereference(it).second
+            result[key] = value
+            postincrement(it)
+        return result, None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_int(CRayStatus status, const optional[int]& c_int):
+    # -> int
+    try:
+        check_status_timeout_as_rpc_error(status)
+        return c_int.value(), None
+    except Exception as e:
+        return None, e
+
+cdef convert_optional_vector_str(
+        CRayStatus status, const optional[c_vector[c_string]]& c_vec):
+    # -> Dict[str, str]
+    try:
+        check_status_timeout_as_rpc_error(status)
+    except Exception as e:
+        return None, e
```

Comment:
why don't we try-catch the whole thing here?

Reply:
because one can't have `cdef` in a try block. Limitations of Cython.

Reply:
nevermind, found out how to do it in try block. updated